rad/prelude/validation.rad

```diff
@@ -58,18 +60,43 @@
         x
         (throw 'validation-failure (string-append "Did not satisfy predicate: " name))))))
 
+(def integral
+  "Validator for whole numbers."
+  (fn [n]
+    (if (integral? n)
+      n
+      (throw 'validation-failure {:info "Number was not an integer."
+                                  :number n}))))
+
 (def key
-  "Given a key and a validator, returns a validator which checks for
-  the existence of that key and that the associated value conforms to the validator."
+  "Given a key `k` and a validator `v`, returns a validator which checks that the
+  value associated to `k` in a dict conforms to `v`. If the key is absent, the
+  validator passes."
   (fn [k v]
     (fn [d]
       (if (member? k d)
         (v (lookup k d))
-        (throw 'validation-failure (string-append "Missing required key: " (show k)))))))
+        d))))
+
+(def contains
+  "Given a value, returns a validator which checks for membership of that value."
+  (fn [k]
+    (fn [x]
+      (if (member? k x)
+        x
+        (throw 'validation-failure {:info "Did not contain key."
+                                    :element k
+                                    :container x})))))
+
+(def has-key
+  "Combines existence and validity of a key in a dict."
+  (fn [k v]
+    (and [(contains k) (key k v)])))
```

Comment:
What do you think about calling this `key` and renaming `key` to `optional-key`? I suspect devs are going to gravitate towards `key` when they actually want `has-key`. I also feel the documentation is a bit to sparse. Can we use the original documentation for `key`?